### ISSUE TITLE
p-table : infinite loop sortFunction with http service

### ISSUE THREAD

**Issue by UlricW** [primefaces/primeng#5163]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => Search github for a similar issue or PR before submitting
[X ] feature request => Please check if request is not on the roadmap already https://github.com/primefaces/primeng/wiki/Roadmap
[ ] support request => Please do not submit support request here, instead see http://forum.primefaces.org/viewforum.php?f=35

**Current behavior**
The sortFunction with an http call trigger an infinite loop:
The sortFunction event is triggered in p-table construction. 
My custom sort function call an http request that return a new array of data with a new reference.
 That reference change trigger the rebuild of the p-table, that recall sortFunction.. the loop is created.

**Expected behavior**
An option to just trigger the sortFunction event on sort icon click

**What is the motivation / use case for changing the behavior?**
My sort is a back-end sort.

* **Angular version:** 5.2.4

* **PrimeNG version:** 5.2.0

* **Browser:** [all]

**Comment by noamichael** [comment]
This can be solved by setting the table to lazy.

HTML:

    <p-table [lazy]="true" sortMode="multiple" [value]="yourValues" (onLazyLoad)="sortFunction($event)"> 
            
    </p-table>

JavaScript:
    
    import { LazyLoadEvent } from 'primeng/api';

    sortFunction($event: LazyLoadEvent) {
        //call your http service here. Read the $event.multiSortMeta or $event.sortField to create your query
    }

**Comment by UlricW** [comment]
Thanks a lot it works. 
An other problem is all columns are sortable, we cannot choose which column are sortable like the dataTable. I think an issue is already open about it.

**Comment by noamichael** [comment]
I think the `pSortableColumn` directive can be used to do this. Here is an example where each column is rendered manually:

    <p-table [lazy]="true" dataKey="id" sortMode="multiple" [value]="values (onLazyLoad)="sort($event)">
            <ng-template pTemplate="header">
                <tr>
                    <th pSortableColumn="entryDate">
                        Entry Date
                      <p-sortIcon field="entryDate"></p-sortIcon>
                    </th>
                    <th>
                        Some Other Column
                    </th>
            </ng-template>
            
    </p-table>

**Comment by UlricW** [comment]
Thanks it works fine. I use a loop to build the template so I put a ternary on the pSortableColumn

**Comment by noamichael** [comment]
@cagataycivici Can this be closed?

**Comment by cagataycivici** [comment]
There is pSortableColumnDisabled property to dusable sorting in dynamic colums.

**Comment by MorlaRamakrishna** [comment]
Same issue occurred in **p-dataTable** how to solve this?

**Comment by kaushikparmar** [comment]
I am getting the same issue.

On sorting, I am calling API. It goes in infinite loop.

This solution given here is not working.

<p-table [value]="rows" [scrollable]="true" [scrollHeight]="300px" [customSort]="true" (sortFunction)="customSort($event)">

**Comment by kaushikparmar** [comment]
Thanks @bakasmarius ,

I am on the way.

Now, I am getting following issues.
- Not able to decide the scroll position from LazyLoadEvent. On up scroll and down scroll, I am getting the same event object.
- when I click on column header for sorting, onLazyLoad event also executed with onSort event.
 
In my case, I have to make http call when scroll ends, to fetch next page data from API.
And On sorting only sort event should be fired not scroll event.
Right now, both events are being executed simultaneously.

        <p-table [value]="row" [scrollable]="true" [scrollHeight]="curentTableHeight" rows="15" [virtualScroll]="true"  [lazy]="true" (onLazyLoad)="loadDataOnScroll($event)" [totalRecords]="totalRecords" [virtualRowHeight]="20" [loading]="isVertualScrollLoading" sortField="merchandise_id" sortMode="single" (onSort)="sortData($event)">

**Comment by bakasmarius** [comment]
I remember I had the same issue. Try removing `(onSort)`. Here's an example of my table that works with lazy loading + sorting:
`      <p-table
        [lazy]="true" [virtualScroll]="true" (onLazyLoad)="displayResults($event);"
        [columns]="tableColumns" [value]="items"
        [resizableColumns]="true" [reorderableColumns]="true" scrollHeight="740px"
        [scrollable]="true" [rows]="lazyLoadIncrement" [virtualRowHeight]="1" [totalRecords]="itemsToLoad"
        [loading]="loading"
      >`

`displayResults()` includes sorting logic + fetching the data from http.
I hope this helps.

**Comment by sagar-kiwi** [comment]
I have query, 
i need to call function on page change and on Sorting , separately,
when i use sortFunction then it is calling infite calls,
please let me know, if there is any solution.

**Comment by kaushikparmar** [comment]
@sagar-kiwi 
Following is my code which may help you solve your problem.

I created custom events/directive (appScroll  & appFillHeight ) and added it to p-table to extend its functionalities as per my requirements.

<app-sort-icon> is a custom directive for customised sorting functionality. I can provide you the code for the same if you want.


<p-table [value]="rows" [sortField]="'accessory_cat_id'" [sortOrder]="sortType" [scrollable]="true" appScroll (scrollDown)="scrollDown()" appFillHeight (tableHeight)="getTableHeight($event)" [scrollHeight]="curentTableHeight" rows="15"  [lazy]="true" (onLazyLoad)="loadDataOnSort($event)" [totalRecords]="totalRecords">
          <ng-template pTemplate="colgroup">
            <colgroup>
              <col style="width: 100px">
              <col style="width: 160px">
              <col style="width: 160px">
              <col style="width: 100px">
              <col style="width: 15em">
            </colgroup>
          </ng-template>
          <ng-template pTemplate="header">
            <tr>
              <th [pSortableColumn]="'accessory_cat_id'">
                <span>Category ID</span>
                <span class="sort-icon">
                  <app-sort-icon [fieldName]="'accessory_cat_id'" [sortEvent]="sortEvent"></app-sort-icon>
                </span>
            </th>
            <th [pSortableColumn]="'name'">
              <span>Display/Category Name</span>
              <span class="sort-icon">
                <app-sort-icon [fieldName]="'name'" [sortEvent]="sortEvent"></app-sort-icon>
              </span>
            </th>
            <th [pSortableColumn]="'name_arabic'">
              <span>Arabic Name</span>
                <span class="sort-icon">
                  <app-sort-icon [fieldName]="'name_arabic'" [sortEvent]="sortEvent"></app-sort-icon>
                </span>
            </th>
            <th [pSortableColumn]="'sequence'">
              <span>Sequence</span>
                <span class="sort-icon">
                  <app-sort-icon [fieldName]="'sequence'" [sortEvent]="sortEvent"></app-sort-icon>
                </span>
            </th>
            <th style="width: 15em">Action</th>
            </tr>
          </ng-template>
          <ng-template pTemplate="body" let-row>
            <tr>
              <td>
                <span>{{row?.accessory_cat_id}}</span>
              </td>
              <td>
                <span *ngIf="!row?.name && !row?.isEditActive">-</span>
                <span *ngIf="row?.name && !row?.isEditActive">{{row?.name}}</span>
                <textarea *ngIf="row?.isEditActive" required maxlength="40" class="form-control input-area"   pattern="^\S+(?: \S+)*$" [ngClass]="{'edit-input': !row?.isEditActive, 'error-border': inlineForm?.controls?.Name?.invalid && (inlineForm?.controls?.Name?.dirty || inlineForm?.controls?.Name?.touched)}" name="Name" [disabled]="!row?.isEditActive" [(ngModel)]="row.name" (ngModelChange)="checkValidations(inlineForm)"></textarea>
              </td>
              <td class="table-rtl">
                <span *ngIf="!row?.name_arabic && !row?.isEditActive">-</span>
                <span *ngIf="row?.name_arabic && !row?.isEditActive">{{row?.name_arabic}}</span>
                <textarea *ngIf="row?.isEditActive" maxlength="40" class="form-control rtl input-area"  [ngClass]="{'edit-input': !row?.isEditActive, 'error-border': inlineForm?.controls?.NameArabic?.invalid && (inlineForm?.controls?.NameArabic?.dirty || inlineForm?.controls?.NameArabic?.touched)}" name="NameArabic" [disabled]="!row?.isEditActive" [(ngModel)]="row.name_arabic" (ngModelChange)="checkValidations(inlineForm)"></textarea>
              </td>
              <td>
                <span *ngIf="!row?.sequence && !row?.isEditActive">-</span>
                <span *ngIf="row?.sequence && !row?.isEditActive">{{row?.sequence}}</span>
                <textarea *ngIf="row?.isEditActive" required maxlength="5" pattern="^[1-9][0-9]*$" class="form-control input-area" [ngClass]="{'edit-input': !row?.isEditActive, 'error-border': inlineForm?.controls?.Sequence?.invalid && (inlineForm?.controls?.Sequence?.dirty || inlineForm?.controls?.Sequence?.touched)}" name="Sequence" [disabled]="!row?.isEditActive" [(ngModel)]="row.sequence" (ngModelChange)="checkValidations(inlineForm)"></textarea>
              </td>
              <td>
                <div class="btn-table-action d-flex" *ngIf="!row?.isEditActive && user_page_roles?.can_update">
                  <button type="button"  class="btn btn-outline-info btn-outline-blue mr-3 edit-btn" (click)="editRow(row)" [disabled]="row?.is_active===false">Edit</button>
                  <button type="button"  class="btn btn-outline-success btn-outline-green mr-3 active-btn" (click)="updateRecord(row, true)" *ngIf="row?.is_active===false">Activate</button>
                  <button type="button"  class="btn btn-outline-success btn-outline-red mr-3 inactive-btn" (click)="updateRecord(row, false)" *ngIf="row?.is_active===true">Deactivate</button>
                </div>
                <div class="btn-table-action d-flex" *ngIf="row?.isEditActive && user_page_roles?.can_update">
                  <button type="button"  class="btn btn-outline-info btn-outline-blue mr-3 edit-btn" (click)="save(row, inlineForm)">Save</button>
                  <button type="button"  class="btn btn-outline-info btn-outline-red mr-3 edit-btn" (click)="cancelAddEditRow(row)">Cancel</button>
                </div>
              </td>
            </tr>
          </ng-template>
          <ng-template pTemplate="emptymessage">
            <tr>
                <td [attr.colspan]="5" class="text-center">
                    <span *ngIf="isRecordFound">No records found</span>
                </td>
            </tr>
          </ng-template>
        </p-table>

**Comment by jayasris** [comment]
Hi there, I am using primeng p-table and have the same infinite loop issue in lazyloading as the columns also get repopulated and that triggers lazyloading continuously. @kaushikparmar How did you solve this issue? Are you able to share the app-sort-icon implementation idea? Thanks!

**Comment by tibinthomas** [comment]
Why this issue is closed? Is there any workarounds for this issue?

**Comment by Ro3A** [comment]
@tibinthomas 

This entire thread is the fix. Depends on your scenario though as to the exact combination of settings to use. I found this issue just now because of the same inifinite loop. I didn't want to use the (onLazyLoad) though because all sorting / paging / filtering for us is done in an API.

So I just set [lazy]=true. Problem solved. You don't have to use the onLazyLoad event if, like me, you are populating the table from an observable that retrieves from http.

For additional context, In my scenario, we have separate components for filtering and it's all event driven. So we have Three behavior subjects. One for FilterOptions, one for PagingOptions, and one for SortOptions. 

I then have a data$ = combineLatest([this._filterOptions, this._pagingOptions, this._sortOptions]).pipe(etc)

We're using [lazy]="true", with (onSort)=onSort($event) which fires the method and in that method we push new values in to the behavior subject. This in turn triggers the combineLatest operator thus updating the main data$ observable, thus updating the table contents.
